MRG: report checksum file download failures

bluegenes opened the pull request:
- Fixes #70 

This PR reports any failures brought about by issues downloading or parsing the md5sum file within `gbsketch`. These cases were not previously included in the failures file. This is the issue in #70.

Now that we have a separate output file, I thought it might be useful to also pull out cases where the md5sum does not match expectation, and write those to the md5sum failure file as well (added in 6755296f80d0a859ecc1a81f73e94418b7bfd120). These were previously just written to the download failures file, with no explanation for _why_ the download failed (e.g. whether it was a download issue or an md5sum matching issue).

There is no straightforward way to test the checksum failures file via `gbsketch`, but we use the mismatch md5sum case to test it via `urlsketch`.

There is a "major" change this PR brings: the `--checksum-fail` file is now required for `gbsketch`. I think it's not possible to make this optional with the current strategy, because otherwise the errors for failed md5sum file download/parsing would not be properly captured anywhere. I have made `--checksum-fail` optional for `urlsketch`, because there folks need to manually provide expected md5sums, so we don't have md5sum file download issues and often we won't even have md5sums to check.

bluegenes commented:
@ccbaumler do you have an accession where the `md5sum` file download or parsing definitely fails? My assumption is that the majority of the issues here happened when NCBI was updating files (and so shouldn't happen most of the time), but I'd like to figure out a way to test the failed checksum file output.

cc @ctb

ctb commented:
looking at the log messages, the checksum file that is being returned is almost certainly some kind of HTML error message.

Hmm, I wonder - are yoweu checking to see if the HTTP status is 200? I would expect the NCBI server to return a non-200 in the case where it's failing for any reason. That might be a simple and easy way to determine if you're getting the expected file...

ccbaumler commented:
An example set from the log file generated during the linked issue:

```
^MESC[K
== This is sourmash version 4.8.9. ==
^MESC[K== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

^MESC[K=> sourmash_plugin_directsketch 0.3.2
^MESC[Kparams: ['dna,k=21,k=31,k=51,scaled=1000,abund']
^MESC[Kdownloading and sketching all accessions in '/group/ctbrowngrp4/2024-ccbaumler-genbank/genbank-20240712/data/missing-genomes.20240712-viral.csv using 5 retries and 4 threads
No protein signature templates provided, and --keep-fasta is not set.
Downloading and sketching genomes only.
Error: Invalid checksum line format in URL https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/023/535/875/GCA_023535875.1_ASM2353587v1/md5checksums.txt: <head>
Error: Invalid checksum line format in URL https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/023/533/615/GCA_023533615.1_ASM2353361v1/md5checksums.txt: <head>
Error: Invalid checksum line format in URL https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/023/533/665/GCA_023533665.1_ASM2353366v1/md5checksums.txt: <head>
```

@ctb had pointed out that `<head>` was most likely from a html return stating either:
- ncbi is down (which happened at least twice recently)
- I over-requested and ncbi refused service :(

bluegenes commented:
Ah, good point @ctb - https://github.com/sourmash-bio/sourmash_plugin_directsketch/pull/92/commits/e883aad829065333510780e032b893135f2adab7 adds a check for the response status for the md5sum file.

I modified the checksum output file to contain the full error. This will provide both the calculated and expected md5sums in the event we have a mismatch. This output file cannot be read in as input to `urlsketch` like the download failures, but I think that's ok because we probably want to 1. try again to download the md5sum file, or 2. investigate what led to the md5sum mismatch.

I've added the failed checksum as optional output to urlsketch, which allows me to test the output format by using the mismatched md5sum case.

bluegenes commented:
ok @ctb this is ready for review!

ctb commented:
thanks @bluegenes - have you had a chance to run it on something that generates the errors? Is this something @ccbaumler could/should do before merge? Or would you prefer merge on the code review and then see what happens with the release?

I could build a wheel for @ccbaumler to install and run out on farm, if you want someone to try "at scale", which is where I think this error shows up.

ctb commented on src/directsketch.rs:
```diff
@@ -115,6 +115,15 @@ async fn download_and_parse_md5(client: &Client, url: &Url) -> Result<HashMap<St
         .context("Failed to send request")?;
 
+    // Check if the status code is 200 OK
+    if !response.status().is_success() {
```
idle curiousity: does `reqwest` allow you to automatically follow HTTP redirects? Is that enabled?

bluegenes replied:
I think it's automatic
> By default, a Client will automatically handle HTTP redirects, having a maximum redirect chain of 10 hops. To customize this behavior, a redirect::Policy can be used with a ClientBuilder.
https://docs.rs/reqwest/latest/reqwest/redirect/index.html#

ctb replied:
fantastic!

ctb commented on src/directsketch.rs:
```diff
@@ -167,7 +176,7 @@ async fn download_with_retry(
                     } else {
                         last_error = Some(anyhow!(
-                            "MD5 hash does not match. Expected: {}, Found: {}",
+                            "MD5 hash does not match. Expected: {}; Found: {}",
```
might be good to put quotes in here?

```suggestion
                            "MD5 hash does not match. Expected: '{}'; Found: '{}'",
```

ctb replied:
oh, never mind, not for an md5. Duh.

bluegenes replied:
good idea, added in https://github.com/sourmash-bio/sourmash_plugin_directsketch/pull/92/commits/139c956a41b5e91d9d41c61ab00d3a55f0c6009a!

bluegenes replied:
whoops, i missed your second comment somehow. I can roll it back if you think it's better without 🤷🏻‍♀️

ctb replied:
oh it's fine there, I just wanted it ALSO in the other error message, `"Invalid checksum line format in URL {}: {}` :)

bluegenes replied:
hah, I should really refresh this page more often when we're doing this in real time. On it 😂

ctb commented on src/directsketch.rs:
```diff
@@ -320,6 +334,33 @@ async fn dl_sketch_assembly_accession(
+                let url = file_type.url(&base_url, &full_name);
+                let file_name = file_type.filename_to_write(&accession);
+                let failed_checksum_download: FailedChecksum = FailedChecksum {
+                    accession: accession.clone(),
```
do these need to be clones?

ctb replied:
I think you might be able to use [field init shorthand syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#using-the-field-init-shorthand), but I'm not 100% sure how the ownership etc thing works here. So, this is a curious question, rather than requesting :)

bluegenes replied:
yes, because I'm using `String`.

If I try to remove the clone:
> move occurs because `accession` has type `std::string::String`, which does not implement the `Copy` trait

I'm open to other solutions, I probably overuse `String`

ctb replied:
right, I'll think on this and see if I can get anywhere :)

bluegenes replied:
if I just use `name` for example instead of `name: name.clone()`, rust-analyzer err:
> consider cloning the value if the performance cost is acceptable: `: name.clone()`

I might be able to avoid the clone by having name be str, not sure

ctb replied:
I don't think so - I think you need to make it clear that the new struct is taking ownership of the old struct fields. Will have to re-read the code and then see if there's a way to tell Rust to cede ownership!

ctb reviewed (approved):
Overall this looks great! I don't quite understand (on a first pass) exactly how you are testing this, but the tests are all there so 👍

bluegenes commented:
> Overall this looks great! I don't quite understand (on a first pass) exactly how you are testing this, but the tests are all there so 👍

thanks! I mean, unfortunately I am _not_ completely testing it within `gbsketch`, because I don't have examples where it will certainly fail. But the same functionality is used across both `gbsketch` and `urlsketch` for checksum failures, so the new struct + output is at least partly tested through the md5sum mismatch test I added for `urlsketch`.

I have not run it at scale -- would be great if @ccbaumler wanted to give it a run.

ctb commented:
@ccbaumler if you wanted to give this a try, the release wheel is at farm:`/group/datalabgrp/ctbrown/sourmash_plugin_directsketch/target/wheels/sourmash_plugin_directsketch-0.3.2-cp312-cp312-manylinux_2_34_x86_64.whl` and you should be able to run something like:
```
module load mamba
mamba create -y -n directsketch_test sourmash python=3.12
mamba activate directsketch_test
pip install /group/datalabgrp/ctbrown/sourmash_plugin_directsketch/target/wheels/sourmash_plugin_directsketch-0.3.2-cp312-cp312-manylinux_2_34_x86_64.whl
```

and then go about your normal business of running directsketch.

bluegenes commented:
@ctb @ccbaumler -- any objects to me merging this one?

ctb commented:
go for it - but if you could create an issue saying that we should try forcing a checksum error somehow to test this better, I'd appreciate it ;)
